_posts/2015-11-03-what-you-can-learn-when-you-pair-and-collaborate.md

```diff
@@ -0,0 +1,201 @@
+                item.increaseQuality();
+            }
+
+            if (item.expiresIn(SIX_DAYS)) {
+                item.increaseQuality();
+            }
+        } else {
+            if (item.name.equals(SULFURAS)) {
+
+            } else {
+                item.decreaseQuality();
+            }
+        }
+
+        // P
+        if (item.name.equals(SULFURAS)) {
+
+        } else {
+            item.decreaseSellIn();
+        }
+
+		// A
+        if (item.name.equals(AGED_BRIE)) {
+            if (item.isExpired()) {
+                item.increaseQuality();
+            }
+        } else {
+        	// B
+            if (item.name.equals(BACKSTAGE_PASSES)) {
+                if (item.isExpired()) {
+                    item.setQualityToMinimum();
+                }
+            } else {
+            	// C
+                if (item.name.equals(SULFURAS)) {
+                    if (item.isExpired()) {
+                        continue;
+                    }
+                } else { // D
+                    if (item.isExpired()) {
+                        item.decreaseQuality();
+                    }
+                }
+            }
+        }
+    }
+}
+```
+
+We can see that the code blocks are not all at the same level of abstraction but also there is duplication across them all. All our tests are green so we can move blocks around without fear as we will be alerted if we break existing functionality. To me the above blocks (O, P, ABCD) appear like the below mathematical expression, and can be merged like this expression:
+
+```P (A + B + C + D) = PA + PB + PC + PD```
+
+but not in this order:
+
+```P (A + B + C + D) = AP + BP + CP + DP``` 
+
+as it won't just work due to the change in order of the blocks, due to the order of execution which is important. In terms of code we get the below compact version, where the order and logic is preserved and duplication is removed (blocks are merged):
+
+```java
+// A
+if (item.name.equals(AGED_BRIE)) {
+    item.decreaseSellIn(); // <==== P
+    if (item.isExpired()) {
+        item.increaseQuality();
+    }
+} else { 
+    // B
+    if (item.name.equals(BACKSTAGE_PASSES)) {
+        item.decreaseSellIn(); // <==== P
+        if (item.isExpired()) {
+            item.setQualityToMinimum();
+        }
+    } else {
+        // C <=== P is not necessary to be applied here
+        if (item.name.equals(SULFURAS)) {
+            if (item.isExpired()) {
+                continue;
+            }
+        } else {
+            // D
+            item.decreaseSellIn(); // <==== P
+            if (item.isExpired()) {
+                item.decreaseQuality();
+            }
+        }
+    }
+}
+```
+
+By doing the above we have brought behaviours that belong to the respective item types together, instead of being in isolation. Now we can merge the remaining two blocks much easier, as they are symmetric in structure. By applying the same principle we saw above, and with that you get the below:
```

Comment:
The two blocks you're mentioning here are O and the new big block `PA + PB + PC + PD` right ? It would be better to mention it.